controllers/phschedule_controller.go

```diff
@@ -17,34 +17,241 @@ package controllers
+			}
+			return ctrl.Result{RequeueAfter: 10 * time.Minute}, nil
+		}
+	} else {
+		recurrence = getDefinedRecurrence(recurType)
+	}
+
+	if recurType == "inactive" {
+		phScheduleCron, ok := r.PhScheduleCronMap[phSchedule.Name]
+		if ok { // if the schedule job has cron already, stop and delete it
+			phScheduleCron.c.Stop()
+			phScheduleCron.c.Remove(phScheduleCron.entryID)
+			delete(r.PhScheduleCronMap, phSchedule.Name)
+		}
+		phSchedule.Status.Valid = true
+		phSchedule.Status.Active = false
+		phSchedule.Status.NextRunTime = nil
+		phSchedule.Status.Message = "This phSchedule is inactive"
+		if err := r.updatePhScheduleStatus(ctx, phSchedule); err != nil {
+			return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
+		}
+		log.Info("the phSchedule is inactive, so delete cron and return")
+		return ctrl.Result{RequeueAfter: 10 * time.Minute}, nil
+	}
+
+	phScheduleCron, ok := r.PhScheduleCronMap[phSchedule.Name]
+	var nextRun metav1.Time
+	var entryID cron.EntryID
+
+	_, err := r.buildPhJob(phSchedule)
+	if err != nil {
+		phSchedule.Status.Valid = false
+		phSchedule.Status.Message = err.Error()
+		phSchedule.Status.NextRunTime = nil
+		phSchedule.Status.Active = true
+		if err := r.updatePhScheduleStatus(ctx, phSchedule); err != nil {
+			return ctrl.Result{RequeueAfter: 1 * time.Minute}, err
+		}
+		return ctrl.Result{RequeueAfter: 10 * time.Minute}, nil
+	}
+
+	createPhJob := func() {
+		log.Info("phSchedule is triggered, create phJob", "phSchedule", phSchedule.Name)
 
-	// your logic here
+		phJob, _ := r.buildPhJob(phSchedule)
+		err = r.Client.Create(ctx, phJob)
 
-	return ctrl.Result{}, nil
+		if err == nil { // create phJob successfully
+			log.Info("cron successfully create phJob", "phJob", phJob.ObjectMeta.Name)
+		} else { // error occurs when creating pod
+			log.Error(err, "cron failed to create phJob", "phJob")
+		}
+
+		log.Info("[In Cron]: after create phjob update nextRunTime", "phSchedule", phSchedule.Name, "recurrence", recurrence, "nextRun", nextRun.String())
+		nextRun = metav1.NewTime(r.PhScheduleCronMap[phSchedule.Name].c.Entry(phScheduleCron.entryID).Next)
+		phSchedule.Status.NextRunTime = &nextRun
+		if err := r.updatePhScheduleStatus(ctx, phSchedule); err != nil {
+		}
+	}
+
+	if !ok { // phSchedule has no cron in controller yet, create one
+		log.Info("phSchedule has no cron, create one")
+		phScheduleCron = &PhScheduleCron{}
+
+		c := cron.New(cron.WithLocation(time.UTC)) // use UTC TimeZone
+		entryID, _ = c.AddFunc(recurrence, createPhJob)
+
+		phScheduleCron.c = c
+		phScheduleCron.entryID = entryID
+		phScheduleCron.recurType = recurType
+		phScheduleCron.recurrence = recurrence
+		phScheduleCron.c.Start()
+
+		r.PhScheduleCronMap[phSchedule.Name] = phScheduleCron
+
+	} else { // phSchedule already has cron in controller, check if the spec changed
+		log.Info("phSchedule has cron, check if the spec changed ")
+		if recurType != phScheduleCron.recurType || recurrence != phScheduleCron.recurrence {
```

Comment:
If the user only edits the command field or change instance type...etc, it will not trigger the following statements. We will be still using old settings.

Reply:
nice catch!
change to overwriting the cron function every reconcile